src/ir/possible-contents.cpp

```diff
@@ -0,0 +1,1607 @@
+      //       we'd need to see which references are used and which escape etc.
+      //       For now, assume a closed world for such such advanced use cases /
+      //       assume this pass won't be run in them anyhow.
+      // TODO: do this only once if multiple tables are exported
+      for (auto& elementSegment : wasm.elementSegments) {
+        for (auto* curr : elementSegment->data) {
+          if (auto* refFunc = curr->dynCast<RefFunc>()) {
+            calledFromOutside(refFunc->func);
+          }
+        }
+      }
+    }
+  }
+
+#ifdef POSSIBLE_CONTENTS_DEBUG
+  std::cout << "func phase\n";
+#endif
+
+  // Connect function parameters to their signature, so that any indirect call
+  // of that signature will reach them.
+  // TODO: find which functions are even taken by reference
+  for (auto& func : wasm.functions) {
+    for (Index i = 0; i < func->getParams().size(); i++) {
+      links.insert(getIndexes({SignatureParamLocation{func->type, i},
+                               LocalLocation{func.get(), i, 0}}));
+    }
+    for (Index i = 0; i < func->getResults().size(); i++) {
+      links.insert(getIndexes({ResultLocation{func.get(), i},
+                               SignatureResultLocation{func->type, i}}));
+    }
+  }
+
+#ifdef POSSIBLE_CONTENTS_DEBUG
+  std::cout << "struct phase\n";
+#endif
+
+  if (getTypeSystem() == TypeSystem::Nominal) {
+    subTypes = std::make_unique<SubTypes>(wasm);
+  }
+
+#ifdef POSSIBLE_CONTENTS_DEBUG
+  std::cout << "Link-targets phase\n";
+#endif
+
+  // Add all links to the targets vectors of the source locations, which we will
+  // use during the flow.
+  for (auto& link : links) {
+    getTargets(link.from).push_back(link.to);
+  }
+
+#ifndef NDEBUG
+  // Each vector of targets (which is a vector for efficiency) must have no
+  // duplicates.
+  for (auto& info : locations) {
+    disallowDuplicates(info.targets);
+  }
+#endif
+
+#ifdef POSSIBLE_CONTENTS_DEBUG
+  std::cout << "roots phase\n";
+#endif
+
+  // Set up the roots, which are the starting state for the flow analysis: send
+  // their initial content to them to start the flow.
+  for (const auto& [location, value] : roots) {
+#if defined(POSSIBLE_CONTENTS_DEBUG) && POSSIBLE_CONTENTS_DEBUG >= 2
+    std::cout << "  init root\n";
+    dump(location);
+    value.dump(std::cout, &wasm);
+    std::cout << '\n';
+#endif
+
+    updateContents(location, value);
+  }
+
+#ifdef POSSIBLE_CONTENTS_DEBUG
+  std::cout << "flow phase\n";
+  size_t iters = 0;
+#endif
+
+  // Flow the data while there is still stuff flowing.
+  while (!workQueue.empty()) {
+#ifdef POSSIBLE_CONTENTS_DEBUG
+    iters++;
+    if ((iters & 255) == 0) {
+      std::cout << iters++ << " iters, work left: " << workQueue.size() << '\n';
+    }
+#endif
+
+    auto iter = workQueue.begin();
+    auto locationIndex = *iter;
+    workQueue.erase(iter);
+
+    flowAfterUpdate(locationIndex);
+
+    updateNewLinks();
```

Comment:
We do all processing from `locationIndex` -> targets in `flowUpdate()` above. Then does adding new links after we've done all the processing anything? (I guess I am misunderstand something.)

Reply:
I'm updating the comment on that method to this, hopefully it is clearer?
```cpp
  // New links must be added to the appropriate |targets| vector (that is, the
  // |targets| of the source must get a new target added to it). We are
  // iterating on |targets| for most of our work, however, so we must not alter
  // it at that time. This method is called later, after there is no risk of
  // aliasing, and it appends to the appropriate |targets| vector.
  void updateNewLinks();
```

Reply:
Yeah the comment looks clearer, but I have another question about this behavior. I asked in in another thread.